cmd/dm-syncer/config.go

```diff
@@ -0,0 +1,357 @@
+	Timezone         *time.Location `json:"-"`
+
+	printVersion bool
+}
+
+// RouteRule is route rule that syncing
+// schema/table to specified schema/table
+// This config has been replaced by `router.TableRule`
+type RouteRule struct {
+	PatternSchema string `toml:"pattern-schema" json:"pattern-schema"`
+	PatternTable  string `toml:"pattern-table" json:"pattern-table"`
+	TargetSchema  string `toml:"target-schema" json:"target-schema"`
+	TargetTable   string `toml:"target-table" json:"target-table"`
+}
+
+// SkipDML defines config rule of skipping dml.
+// This config has been replaced by BinLog filter.
+type SkipDML struct {
+	Schema string `toml:"db-name" json:"db-name"`
+	Table  string `toml:"tbl-name" json:"tbl-name"`
+	Type   string `toml:"type" json:"type"`
+}
+
+func (oc *oldConfig) convertToNewFormat() (*config.SubTaskConfig, error) {
+
+	newTask := &config.SubTaskConfig{
+		LogLevel:  oc.LogLevel,
+		LogFile:   oc.LogFile,
+		LogRotate: oc.LogRotate,
+
+		StatusAddr: oc.StatusAddr,
+		ServerID:   uint32(oc.ServerID),
+		Flavor:     oc.Flavor,
+
+		SyncerConfig: config.SyncerConfig{
+			MetaFile:         oc.Meta,
+			WorkerCount:      oc.WorkerCount,
+			Batch:            oc.Batch,
+			MaxRetry:         oc.MaxRetry,
+			AutoFixGTID:      oc.AutoFixGTID,
+			EnableGTID:       oc.EnableGTID,
+			EnableANSIQuotes: oc.EnableANSIQuotes,
+			DisableCausality: oc.DisableCausality,
+			SafeMode:         oc.SafeMode,
+		},
+
+		BWList: &filter.Rules{
+			DoTables:     oc.DoTables,
+			DoDBs:        oc.DoDBs,
+			IgnoreTables: oc.IgnoreTables,
+			IgnoreDBs:    oc.IgnoreDBs,
+		},
+
+		ConfigFile: oc.ConfigFile,
+		Timezone:   oc.TimezoneStr,
+		From:       oc.From,
+		To:         oc.To,
+	}
+
+	for _, rule := range oc.RouteRules {
+		newTask.RouteRules = append(newTask.RouteRules, &router.TableRule{
+			SchemaPattern: rule.PatternSchema,
+			TablePattern:  rule.PatternTable,
+			TargetSchema:  rule.TargetSchema,
+			TargetTable:   rule.TargetTable,
+		})
+	}
+
+	var err error
+	newTask.FilterRules, err = generateBinlogEventRule(oc.SkipDDLs, oc.SkipDMLs)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	err = newTask.Adjust()
+	return newTask, err
+}
+
+func generateBinlogEventRule(skipDDLs []string, skipDMLs []*SkipDML) ([]*bf.BinlogEventRule, error) {
+	result := make([]*bf.BinlogEventRule, 0, 1+len(skipDMLs))
+	ddlEvents := &bf.BinlogEventRule{}
+	for _, skipDDL := range skipDDLs {
+		if tp, _ := bf.ClassifyEvent(bf.EventType(skipDDL)); tp != "ddl" {
+			return nil, errors.NotValidf("event type %s", skipDDL)
+		}
+		ddlEvents.Events = append(ddlEvents.Events, bf.EventType(skipDDL))
+	}
+	for _, skipDML := range skipDMLs {
+		if tp, _ := bf.ClassifyEvent(bf.EventType(skipDML.Type)); tp != "dml" {
+			return nil, errors.NotValidf("event type %s", skipDML)
+		}
+		result = append(result, &bf.BinlogEventRule{
+			SchemaPattern: skipDML.Schema,
+			TablePattern:  skipDML.Table,
+			Events:        []bf.EventType{bf.EventType(skipDML.Type)},
```

Comment:
@hanfei1991 friendly ping.

Reply:
I have checked it by 
```go
if tp, _ := bf.ClassifyEvent(bf.EventType(skipDML.Type)); tp != "dml" {
		return nil, errors.NotValidf("event type %s", skipDML.Type)
}
```
Is it ok ?

Reply:
It may have problems if there are multi rules with the same `db-name` and `tbl-name`:
```
[[skip-dmls]]
db-name = "foo"
tbl-name = "bar"
type = "delete"

[[skip-dmls]]
db-name = "foo"
tbl-name = "bar"
type = "insert"
```

because in _binlog filter_, it uses the schema name and table name to construct the trie tree and will report an error if matched an existing rule, see https://github.com/pingcap/tidb-tools/blob/11fb19bcad8c68b8ebf6083d7fffd5789bb2a17e/pkg/table-rule-selector/trie_selector.go#L348.

Reply:
I know, so I should merge events with some db-table name